File: common/src/main/java/org/sqlunet/history/HistoryFragment.java
/*
 * Copyright (c) 2019-2023. Bernard Bou
 */

package org.sqlunet.history;

import android.content.Context;
import android.content.Intent;
import android.content.SearchRecentSuggestionsProvider;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.os.ParcelFileDescriptor;
import android.util.Log;
import android.view.GestureDetector;
import android.view.GestureDetector.SimpleOnGestureListener;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.CursorAdapter;
import android.widget.ListView;
import android.widget.SimpleCursorAdapter;
import android.widget.Toast;

import org.sqlunet.browser.common.R;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.Reader;
import java.io.Writer;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContract;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.loader.app.LoaderManager;
import androidx.loader.app.LoaderManager.LoaderCallbacks;
import androidx.loader.content.Loader;

/**
 * History activity
 *
 * @author Bernard Bou
 * @noinspection WeakerAccess
 */
public class HistoryFragment extends Fragment implements LoaderCallbacks<Cursor>, AdapterView.OnItemClickListener
{
	private static final String TAG = "HistoryF";

	/**
	 * Export/import text file
	 */
	private static final String HISTORY_FILE = "semantikos_search_history.txt";

	/**
	 * Cursor loader id
	 */
	private static final int LOADER_ID = 2222;

	/**
	 * List view
	 */
	private ListView listView;

	/**
	 * Cursor adapter
	 */
	private CursorAdapter adapter;

	@Override
	public void onCreate(@Nullable final Bundle savedInstanceState)
	{
		super.onCreate(savedInstanceState);

		// launchers
		registerLaunchers();

		setHasOptionsMenu(true);
	}

	@Override
	public View onCreateView(@NonNull final LayoutInflater inflater, final ViewGroup container, @Nullable final Bundle savedInstanceState)
	{
		final View view = inflater.inflate(R.layout.fragment_history, container, false);

		// list adapter bound to the cursor
		this.adapter = new SimpleCursorAdapter(this, // context
				R.layout.item_history, // row template to use android.R.layout.simple_list_item_1
				null, // empty cursor to bind to
				new String[]{SearchRecentSuggestions.SuggestionColumns.DISPLAY1}, // cursor columns to bind to
				new int[]{android.R.id.text1}, // objects to bind to those columns
				0);

		// list view
		this.listView = view.findViewById(android.R.id.list);

		// bind to adapter
		this.listView.setAdapter(this.adapter);

		// click listener
		this.listView.setOnItemClickListener(this);

		// swipe
		final SwipeGestureListener gestureListener = new SwipeGestureListener();
		this.listView.setOnTouchListener(gestureListener);

		// initializes the cursor loader
		LoaderManager.getInstance(this).initLoader(HistoryFragment.LOADER_ID, null, this);

		return view;
	}


	@SuppressWarnings("SameReturnValue")
	@Override
	public boolean onCreateOptionsMenu(@NonNull final Menu menu, @NonNull final MenuInflater inflater)
	{
		inflater.inflate(R.menu.history, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(@NonNull final MenuItem item)
	{
		int itemId = item.getItemId();
		if (itemId == R.id.action_history_export)
		{
			exportHistory();
		}
		else if (itemId == R.id.action_history_import)
		{
			importHistory();
		}
		else if (itemId == R.id.action_history_clear)
		{
			final android.provider.SearchRecentSuggestions suggestions = new android.provider.SearchRecentSuggestions(requireContext(), SearchRecentSuggestions.getAuthority(requireContext()), SearchRecentSuggestionsProvider.DATABASE_MODE_QUERIES);
			suggestions.clearHistory();
			return true;
		}
		return false;
	}

	@NonNull
	@Override
	public Loader<Cursor> onCreateLoader(final int loaderID, final Bundle args)
	{
		// assert loaderID == LOADER_ID;
		final SearchRecentSuggestions suggestions = new SearchRecentSuggestions(requireContext(), SearchRecentSuggestionsProvider.DATABASE_MODE_QUERIES);
		return suggestions.cursorLoader();
	}

	@Override
	public void onLoadFinished(@NonNull final Loader<Cursor> loader, @NonNull final Cursor cursor)
	{
		cursor.moveToFirst();
		this.adapter.swapCursor(cursor);
	}

	@Override
	public void onLoaderReset(@NonNull final Loader<Cursor> loader)
	{
		//noinspection EmptyTryBlock
		try (Cursor ignored = this.adapter.swapCursor(null))
		{
		}
	}

	// C L I C K

	@Override
	public void onItemClick(final AdapterView<?> parent, final View view, final int position, final long id)
	{
		Log.d(TAG, "Select " + position);
		final Cursor cursor = ((SimpleCursorAdapter) this.listView.getAdapter()).getCursor();
		cursor.moveToPosition(position);
		if (!cursor.isAfterLast())
		{
			final int dataIdx = cursor.getColumnIndex(SearchRecentSuggestions.SuggestionColumns.DISPLAY1);
			assert dataIdx != -1;
			final String query = cursor.getString(dataIdx);
			if (null != query)
			{
				final Intent intent = History.makeSearchIntent(requireContext(), query);
				startActivity(intent);
			}
		}
	}

	// S W I P E

	private class SwipeGestureListener extends SimpleOnGestureListener implements OnTouchListener
	{
		static final int SWIPE_MIN_DISTANCE = 120;
		static final int SWIPE_MAX_OFF_PATH = 250;
		static final int SWIPE_THRESHOLD_VELOCITY = 200;

		@NonNull
		private final GestureDetector gestureDetector;

		@SuppressWarnings("WeakerAccess")
		public SwipeGestureListener()
		{
			this.gestureDetector = new GestureDetector(requireContext(), this);
		}

		@Override
		public boolean onFling(@Nullable final MotionEvent e1, @NonNull final MotionEvent e2, final float velocityX, final float velocityY)
		{
			if (e1 == null)
			{
				return false;
			}

			final int position = HistoryFragment.this.listView.pointToPosition(Math.round(e1.getX()), Math.round(e1.getY()));

			if (Math.abs(e1.getY() - e2.getY()) <= SwipeGestureListener.SWIPE_MAX_OFF_PATH)
			{
				if (Math.abs(velocityX) >= SwipeGestureListener.SWIPE_THRESHOLD_VELOCITY)
				{
					if (e2.getX() - e1.getX() > SwipeGestureListener.SWIPE_MIN_DISTANCE)
					{
						// final SimpleCursorAdapter adapter = (SimpleCursorAdapter) listView.getAdapter();
						final Cursor cursor = HistoryFragment.this.adapter.getCursor();
						if (!cursor.isAfterLast())
						{
							if (cursor.moveToPosition(position))
							{
								final int itemIdIdx = cursor.getColumnIndex("_id");
								assert itemIdIdx != -1;
								final String itemId = cursor.getString(itemIdIdx);

								final int dataIdx = cursor.getColumnIndex(SearchRecentSuggestions.SuggestionColumns.DISPLAY1);
								assert dataIdx != -1;
								final String data = cursor.getString(dataIdx);

								final SearchRecentSuggestions suggestions = new SearchRecentSuggestions(requireContext(), SearchRecentSuggestionsProvider.DATABASE_MODE_QUERIES);
								suggestions.delete(itemId);
								final Cursor cursor2 = suggestions.cursor();
								HistoryFragment.this.adapter.swapCursor(cursor2);

								Toast.makeText(requireContext(), getResources().getString(R.string.title_history_deleted) + ' ' + data, Toast.LENGTH_SHORT).show();
								return true;
							}
						}
						// do not close: cursor.close();
					}
				}
			}

			return super.onFling(e1, e2, velocityX, velocityY);
		}

		@Override
		public boolean onTouch(@NonNull final View view, @NonNull final MotionEvent event)
		{
			if (event.getAction() == MotionEvent.ACTION_UP)
			{
				view.performClick();
			}
			return this.gestureDetector.onTouchEvent(event);
		}
	}

	// I M P O R T / E X P O R T

	private static final String MIME_TYPE = "text/plain";

	/**
	 * Export history
	 */
	private void exportHistory()
	{
		exportLauncher.launch(MIME_TYPE);
	}

	/**
	 * Import history
	 */
	private void importHistory()
	{
		importLauncher.launch(new String[]{MIME_TYPE});
	}

	// D O C U M E N T   I N T E R F A C E

	// You can do the assignment inside onAttach or onCreate, i.e, before the activity is displayed
	private ActivityResultLauncher<String> exportLauncher;

	private ActivityResultLauncher<String[]> importLauncher;

	private void registerLaunchers()
	{
		final ActivityResultContract<String, Uri> createContract = new ActivityResultContracts.CreateDocument(MIME_TYPE)
		{
			@NonNull
			@Override
			public Intent createIntent(@NonNull final Context context, @NonNull final String input)
			{
				final Intent intent = super.createIntent(context, input);
				intent.putExtra(Intent.EXTRA_TITLE, HISTORY_FILE);
				//intent.putExtra(DocumentsContract.EXTRA_INITIAL_URI, pickerInitialUri);
				return intent;
			}
		};
		this.exportLauncher = registerForActivityResult(createContract, uri -> {

			// The result data contains a URI for the document or directory that the user selected.
			if (uri != null)
			{
				doExportHistory(uri);
			}
		});

		final ActivityResultContract<String[], Uri> openContract = new ActivityResultContracts.OpenDocument()
		{
			@NonNull
			@Override
			public Intent createIntent(@NonNull final Context context, @NonNull final String[] input)
			{
				final Intent intent = super.createIntent(context, input);
				intent.putExtra(Intent.EXTRA_TITLE, HISTORY_FILE);
				//intent.putExtra(DocumentsContract.EXTRA_INITIAL_URI, pickerInitialUri);
				return intent;
			}
		};
		this.importLauncher = registerForActivityResult(openContract, uri -> {

			// The result data contains a URIs for the document or directory that the user selected.
			if (uri != null)
			{
				doImportHistory(uri);
			}
		});
	}

	/**
	 * Export history
	 */
	private void doExportHistory(@NonNull final Uri uri)
	{
		Log.d(HistoryFragment.TAG, "Exporting to " + uri);
		try ( //
		      final ParcelFileDescriptor pfd = requireContext().getContentResolver().openFileDescriptor(uri, "w") //
		)
		{
			assert pfd != null;
			try (final OutputStream fileOutputStream = new FileOutputStream(pfd.getFileDescriptor());//
			     final Writer writer = new OutputStreamWriter(fileOutputStream);//
			     final BufferedWriter bw = new BufferedWriter(writer))
			{
				final SearchRecentSuggestions suggestions = new SearchRecentSuggestions(requireContext(), SearchRecentSuggestionsProvider.DATABASE_MODE_QUERIES);
				final Cursor cursor = suggestions.cursor();
				assert cursor != null;
				if (cursor.moveToFirst())
				{
					do
					{
						final int dataIdx = cursor.getColumnIndex(SearchRecentSuggestions.SuggestionColumns.DISPLAY1);
						assert dataIdx != -1;
						final String data = cursor.getString(dataIdx);
						bw.write(data + '\n');
					}
					while (cursor.moveToNext());
				}
				cursor.close();
				Log.i(HistoryFragment.TAG, "Exported to " + uri);
				Toast.makeText(requireContext(), getResources().getText(R.string.title_history_export) + " " + uri, Toast.LENGTH_SHORT).show();
			}
		}
		catch (@NonNull final IOException e)
		{
			Log.e(HistoryFragment.TAG, "While writing", e);
			Toast.makeText(requireContext(), getResources().getText(R.string.error_export) + " " + uri, Toast.LENGTH_SHORT).show();
		}
	}

	/**
	 * Import history
	 */
	private void doImportHistory(@NonNull final Uri uri)
	{
		Log.d(HistoryFragment.TAG, "Importing from " + uri);
		try ( //
		      final InputStream is = requireContext().getContentResolver().openInputStream(uri); //
		      final Reader reader = new InputStreamReader(is); //
		      final BufferedReader br = new BufferedReader(reader) //
		)
		{
			String line;
			while ((line = br.readLine()) != null)
			{
				History.recordQuery(requireContext(), line.trim());
			}
			Log.i(HistoryFragment.TAG, "Imported from " + uri);
			Toast.makeText(requireContext(), getResources().getText(R.string.title_history_import) + " " + uri, Toast.LENGTH_SHORT).show();
		}
		catch (@NonNull final IOException e)
		{
			Log.e(HistoryFragment.TAG, "While reading", e);
			Toast.makeText(requireContext(), getResources().getText(R.string.error_import) + " " + uri, Toast.LENGTH_SHORT).show();
		}
	}
}

File: common/src/main/java/org/sqlunet/browser/history/HistoryFragment.kt
/*
 * Copyright (c) 2025. Bernard Bou <1313ou@gmail.com>
 */
package org.sqlunet.browser.history

import android.annotation.SuppressLint
import android.content.Context
import android.content.SearchRecentSuggestionsProvider
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.result.contract.ActivityResultContracts.CreateDocument
import androidx.core.view.MenuHost
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.loader.app.LoaderManager
import androidx.loader.content.Loader
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.RecyclerView
import org.sqlunet.browser.AppContext
import org.sqlunet.browser.common.R
import org.sqlunet.browser.history.History.makeSearchIntent
import org.sqlunet.browser.history.History.recordQuery
import org.sqlunet.browser.history.SearchRecentSuggestions.Companion.getAuthority
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.FileOutputStream
import java.io.IOException
import java.io.InputStreamReader
import java.io.OutputStreamWriter

class HistoryFragment : Fragment(), LoaderManager.LoaderCallbacks<Cursor> {

    private lateinit var recyclerView: RecyclerView
    private lateinit var adapter: HistoryAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        registerLaunchers()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_history, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        recyclerView = view.findViewById(R.id.recycler_view)
        adapter = HistoryAdapter(requireContext(), null)
        recyclerView.adapter = adapter

        val swipeCallback = object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder): Boolean {
                return false
            }

            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.bindingAdapterPosition
                val cursor = adapter.getCursor()
                if (cursor != null && cursor.moveToPosition(position)) {
                    val itemIdIdx = cursor.getColumnIndex("_id")
                    val itemId = cursor.getString(itemIdIdx)
                    val dataIdx = cursor.getColumnIndex(SearchRecentSuggestions.SuggestionColumns.DISPLAY1)
                    val data = cursor.getString(dataIdx)
                    val suggestions = SearchRecentSuggestions(AppContext.context, SearchRecentSuggestionsProvider.DATABASE_MODE_QUERIES)
                    suggestions.delete(itemId)
                    // Restart the loader to get the updated cursor
                    LoaderManager.getInstance(this@HistoryFragment).restartLoader(LOADER_ID, null, this@HistoryFragment)
                    Toast.makeText(requireContext(), resources.getString(R.string.title_history_deleted) + ' ' + data, Toast.LENGTH_SHORT).show()
                }
            }
        }
        val touchHelper = ItemTouchHelper(swipeCallback)
        touchHelper.attachToRecyclerView(recyclerView)

        val menuHost: MenuHost = requireActivity()
        menuHost.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.history, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                when (menuItem.itemId) {
                    R.id.action_history_export -> exportHistory()
                    R.id.action_history_import -> importHistory()
                    R.id.action_history_clear -> {
                        val suggestions = android.provider.SearchRecentSuggestions(AppContext.context, getAuthority(AppContext.context), SearchRecentSuggestionsProvider.DATABASE_MODE_QUERIES)
                        suggestions.clearHistory()
                        return true
                    }
                }
                return false
            }
        }, getViewLifecycleOwner(), Lifecycle.State.RESUMED)
    }

    override fun onStart() {
        super.onStart()
        LoaderManager.getInstance(this).initLoader(LOADER_ID, null, this)
    }

    override fun onDestroy() {
        super.onDestroy()
        if (::exportLauncher.isInitialized) {
            exportLauncher.unregister()
        }
        if (::importLauncher.isInitialized) {
            importLauncher.unregister()
        }
    }

    override fun onCreateLoader(loaderID: Int, args: Bundle?): Loader<Cursor> {
        val suggestions = SearchRecentSuggestions(AppContext.context, SearchRecentSuggestionsProvider.DATABASE_MODE_QUERIES)
        return suggestions.cursorLoader()
    }

    override fun onLoadFinished(loader: Loader<Cursor>, cursor: Cursor) {
        adapter.changeCursor(cursor)
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {
        adapter.changeCursor(null)
    }

    class HistoryAdapter(private val context: Context, private var cursor: Cursor?) : RecyclerView.Adapter<HistoryAdapter.ViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.item_history, parent, false)
            return ViewHolder(view)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            if (cursor?.moveToPosition(position) == true) {
                val dataIdx = cursor!!.getColumnIndex(SearchRecentSuggestions.SuggestionColumns.DISPLAY1)
                val query = cursor!!.getString(dataIdx)
                holder.textView.text = query
                holder.itemView.setOnClickListener {
                    val intent = makeSearchIntent(context, query)
                    context.startActivity(intent)
                }
            }
        }

        override fun getItemCount(): Int {
            return cursor?.count ?: 0
        }

        @SuppressLint("NotifyDataSetChanged")
        fun changeCursor(newCursor: Cursor?) {
            if (cursor === newCursor) {
                return
            }
            cursor = newCursor
            notifyDataSetChanged()
        }

        fun getCursor(): Cursor? {
            return cursor
        }

        class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
            val textView: TextView = itemView.findViewById(android.R.id.text1)
        }
    }

    // E X P O R T

    private fun exportHistory() {
        exportLauncher.launch(MIME_TYPE)
    }

    private fun importHistory() {
        importLauncher.launch(arrayOf(MIME_TYPE))
    }

    private lateinit var exportLauncher: ActivityResultLauncher<String>

    private lateinit var importLauncher: ActivityResultLauncher<Array<String>>

    private fun registerLaunchers() {
        exportLauncher = registerForActivityResult(CreateDocument(MIME_TYPE)) { uri: Uri? ->
            uri?.let { doExportHistory(it) }
        }
        importLauncher = registerForActivityResult(ActivityResultContracts.OpenDocument()) { uri: Uri? ->
            uri?.let { doImportHistory(it) }
        }
    }

    private fun doExportHistory(uri: Uri) {
        Log.d(TAG, "Exporting to $uri")
        try {
            AppContext.context.contentResolver.openFileDescriptor(uri, "w").use { pfd ->
                FileOutputStream(pfd!!.fileDescriptor).use { fileOutputStream ->
                    OutputStreamWriter(fileOutputStream).use { writer ->
                        BufferedWriter(writer).use { bw ->
                            val suggestions = SearchRecentSuggestions(AppContext.context, SearchRecentSuggestionsProvider.DATABASE_MODE_QUERIES)
                            suggestions.cursor().use { cursor ->
                                if (cursor?.moveToFirst() == true) {
                                    do {
                                        val dataIdx = cursor.getColumnIndex(SearchRecentSuggestions.SuggestionColumns.DISPLAY1)
                                        val data = cursor.getString(dataIdx)
                                        bw.write(data + '\n')
                                    } while (cursor.moveToNext())
                                }
                            }
                            Log.i(TAG, "Exported to $uri")
                            Toast.makeText(requireContext(), resources.getText(R.string.title_history_export).toString() + " " + uri, Toast.LENGTH_SHORT).show()
                        }
                    }
                }
            }
        } catch (e: IOException) {
            Log.e(TAG, "While writing", e)
            Toast.makeText(requireContext(), resources.getText(R.string.error_export).toString() + " " + uri, Toast.LENGTH_SHORT).show()
        }
    }

    private fun doImportHistory(uri: Uri) {
        Log.d(TAG, "Importing from $uri")
        try {
            AppContext.context.contentResolver.openInputStream(uri).use { `is` ->
                InputStreamReader(`is`).use { reader ->
                    BufferedReader(reader).use { br ->
                        var line: String?
                        while (br.readLine().also { line = it } != null) {
                            recordQuery(AppContext.context, line!!.trim { it <= ' ' })
                        }
                        Log.i(TAG, "Imported from $uri")
                        Toast.makeText(requireContext(), resources.getText(R.string.title_history_import).toString() + " " + uri, Toast.LENGTH_SHORT).show()
                    }
                }
            }
        } catch (e: IOException) {
            Log.e(TAG, "While reading", e)
            Toast.makeText(requireContext(), resources.getText(R.string.error_import).toString() + " " + uri, Toast.LENGTH_SHORT).show()
        }
    }

    companion object {
        private const val TAG = "HistoryF"

        private const val LOADER_ID = 2222

        private const val MIME_TYPE = "text/plain"

        // private const val HISTORY_FILE = "semantikos_search_history.txt"
    }
}